pandas/core/categorical.py

```diff
@@ -2279,7 +2279,7 @@ def _recode_for_categories(codes, old_categories, new_categories):
 
     if len(old_categories) == 0:
         # All null anyway, so just retain the nulls
-        return codes
+        return codes.copy()
```

Comment:
`return codes` causes writable flag to be `False` hence the error reported in #18051